src/spatial/utils.jl

```diff
@@ -0,0 +1,209 @@
+"""
+A file with types, structs and functions for spatial simulations
+"""
+
+"""
+stores info for a spatial jump
+"""
+struct SpatialJump{J}
+    src::J    # source location
+    jidx::Int # index of jump as a diffusion hop or reaction
+    dst::J    # destination location
+end
+
+############ abstract spatial system struct ##################
+"""
+Contains all info about the topology of the system
+"""
+abstract type AbstractSpatialSystem end
+
+"""
+returns total number of sites
+"""
+function num_sites end
+
+"""
+returns neighbors of site
+"""
+function neighbors end
+
+################### LightGraph ########################
+num_sites(graph) = nv(graph)
+
+################### CartesianGrid <: AbstractSpatialSystem ########################
+"""
+Cartesian Grid of dimension D
+"""
+struct CartesianGrid{D} <: AbstractSpatialSystem
+    linear_sizes::Vector{Int} #side lengths of the grid
+end
+
+function CartesianGrid(linear_sizes::Vector)
+    CartesianGrid{length(linear_sizes)}(linear_sizes)
+end
+
+function CartesianGrid(dimension, linear_size)
+    CartesianGrid{dimension}([linear_size for i in 1:dimension])
+end
+
+dimension(grid::CartesianGrid{D}) where D = D
+num_sites(grid::CartesianGrid) = prod(grid.linear_sizes)
+
+to_coordinates(grid::CartesianGrid{1}, site) = site
+to_coordinates(grid::CartesianGrid{2}, site) = (mod1(site, grid.linear_sizes[1]),fld1(site, grid.linear_sizes[1]))
+function to_coordinates(grid::CartesianGrid{3}, site)
+    temp = mod1(site,grid.linear_sizes[1]*grid.linear_sizes[2])
+    (mod1(temp, grid.linear_sizes[1]),fld1(temp, grid.linear_sizes[1]), fld1(site, grid.linear_sizes[1]*grid.linear_sizes[2]))
+end
+
+from_coordinates(grid::CartesianGrid{1}, x) = x
+from_coordinates(grid::CartesianGrid{2}, (x,y)) = (y-1) * grid.linear_sizes[1] + x
+from_coordinates(grid::CartesianGrid{3}, (x,y,z)) = (y-1) * grid.linear_sizes[1] + x + (z-1)*grid.linear_sizes[1]*grid.linear_sizes[2]
+
+is_site(grid,site_id::Int) = site_id >= 1 && site_id <= num_sites(grid)
+function is_site(grid,site_coordinates::Tuple)
+    length(site_coordinates) == dimension(grid) || return false
+    for (i,c) in enumerate(site_coordinates)
+        1 <= c && c <= grid.linear_sizes[i] || return false
+    end
+    return true
+end
+
+# TODO make these iterators so they don't allocate memory
```

Comment:
Iterators would be better, but even a Tuple of CartesianIndices would be an improvement.

Reply:
Will do after the PR is merged as agreed today